contracts/src/v0.8/dev/automation/2_1/AutomationRegistrar2_1.sol

```diff
@@ -0,0 +1,518 @@
+   * @member approvedCount the count of upkeeps auto approved of this type
+   */
+  struct UpkeepRegistrationState {
+    AutoApproveType autoApproveType;
+    uint32 autoApproveMaxAllowed;
+    uint32 approvedCount;
+  }
+
+  /**
+   * @notice UpkeepTypeConfig configures the auto-approval levels for upkeeps by type
+   * @member autoApproveType the auto approval setting (see enum)
+   * @member autoApproveMaxAllowed the max number of upkeeps that can be auto approved of this type
+   */
+  struct UpkeepTypeConfig {
+    uint8 upkeepType;
+    AutoApproveType autoApproveType;
+    uint32 autoApproveMaxAllowed;
+  }
+
+  struct RegistrarConfig {
+    IKeeperRegistryMaster keeperRegistry;
+    uint96 minLINKJuels;
+  }
+
+  struct PendingRequest {
+    address admin;
+    uint96 balance;
+  }
+
+  struct RegistrationParams {
+    string name;
+    bytes encryptedEmail;
+    address upkeepContract;
+    uint32 gasLimit;
+    address adminAddress;
+    uint8 triggerType;
+    bytes checkData;
+    bytes triggerConfig;
+    bytes offchainConfig;
+    uint96 amount;
+  }
+
+  RegistrarConfig private s_config;
+  // Only applicable if s_config.configType is ENABLED_SENDER_ALLOWLIST
+  mapping(address => bool) private s_autoApproveAllowedSenders;
+
+  event RegistrationRequested(
+    bytes32 indexed hash,
+    string name,
+    bytes encryptedEmail,
+    address indexed upkeepContract,
+    uint32 gasLimit,
+    address adminAddress,
+    uint8 triggerType,
+    bytes triggerConfig,
+    bytes checkData,
+    uint96 amount
+  );
+
+  event RegistrationApproved(bytes32 indexed hash, string displayName, uint256 indexed upkeepId);
+
+  event RegistrationRejected(bytes32 indexed hash);
+
+  event AutoApproveAllowedSenderSet(address indexed senderAddress, bool allowed);
+
+  event ConfigChanged(address keeperRegistry, uint96 minLINKJuels);
+
+  error InvalidAdminAddress();
+  error RequestNotFound();
+  error HashMismatch();
+  error OnlyAdminOrOwner();
+  error InsufficientPayment();
+  error RegistrationRequestFailed();
+  error OnlyLink();
+  error AmountMismatch();
+  error SenderMismatch();
+  error FunctionNotPermitted();
+  error LinkTransferFailed(address to);
+  error InvalidDataLength();
+
+  /*
+   * @param LINKAddress Address of Link token
+   * @param autoApproveType setting for auto-approve registrations
+   * @param autoApproveMaxAllowed max number of registrations that can be auto approved
+   * @param keeperRegistry keeper registry address
+   * @param minLINKJuels minimum LINK that new registrations should fund their upkeep with
+   */
+  constructor(
+    address LINKAddress,
+    address keeperRegistry,
+    uint96 minLINKJuels,
+    UpkeepTypeConfig[] memory upkeepConfigs
+  ) ConfirmedOwner(msg.sender) {
+    LINK = LinkTokenInterface(LINKAddress);
+    setRegistrationConfig(keeperRegistry, minLINKJuels);
+    for (uint256 idx = 0; idx < upkeepConfigs.length; idx++) {
+      setUpkeepTypeConfig(upkeepConfigs[idx]);
```

Comment:
is it generally ok to call external functions on the contract? (I have mostly only seen internal functions being called)

Reply:
yeah I think this is super common. That's the point of the `public` visibility declaration - a func for both external and internal use.